core/test/matrix/csr_builder.cpp

```diff
@@ -0,0 +1,96 @@
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include "core/matrix/csr_builder.hpp"
+
+
+#include <memory>
+
+
+#include <gtest/gtest.h>
+
+
+namespace {
+
+
+class CsrBuilder : public ::testing::Test {
+protected:
+    using Mtx = gko::matrix::Csr<>;
+
+    CsrBuilder()
+        : exec(gko::ReferenceExecutor::create()),
+          mtx(Mtx::create(exec, gko::dim<2>{2, 3}, 4))
+    {}
+
+    std::shared_ptr<const gko::Executor> exec;
+    std::unique_ptr<Mtx> mtx;
+};
+
+
+TEST_F(CsrBuilder, ReturnsCorrectArrays)
+{
+    gko::matrix::CsrBuilder<> builder{mtx.get()};
+
+    auto builder_col_idxs = builder.get_col_idx_array().get_data();
+    auto builder_values = builder.get_value_array().get_data();
+    auto ref_col_idxs = mtx->get_col_idxs();
+    auto ref_values = mtx->get_values();
+
+    ASSERT_EQ(builder_col_idxs, ref_col_idxs);
+    ASSERT_EQ(builder_values, ref_values);
+}
+
+
+TEST_F(CsrBuilder, UpdatesSrowOnDestruction)
+{
+    struct mock_strategy : public Mtx::strategy_type {
+        virtual void process(const gko::Array<gko::int32> &,
+                             gko::Array<gko::int32> *)
+        {
+            *was_called = true;
+        }
+        virtual int64_t clac_size(const int64_t nnz) { return 0; }
+
+        mock_strategy(bool &flag) : Mtx::strategy_type(""), was_called(&flag) {}
+
+        bool *was_called;
+    };
+    bool was_called{};
+    mtx->set_strategy(std::make_shared<mock_strategy>(was_called));
+
+    gko::matrix::CsrBuilder<>{mtx.get()};
```

Comment:
Is `was_called` supposed to be set on the destruction of this object? In that case, why does it work currently, shouldn't this be scoped inside a `{}` block? Currently, the destruction should happen at the very end of this block, after the `ASSERT_TRUE`.

Reply:
I think what you're testing is not destruction related, but the call to `make_srow` in `set_strategy` (which calls process further down).

Reply:
> Is `was_called` supposed to be set on the destruction of this object? In that case, why does it work currently, shouldn't this be scoped inside a `{}` block? Currently, the destruction should happen at the very end of this block, after the `ASSERT_TRUE`.

The CsrBuilder is a temporary, as it doesn't have a variable name. Thus it get destroyed immediately.


> I think what you're testing is not destruction related, but the call to `make_srow` in `set_strategy` (which calls process further down). Maybe I misunderstood the test name.

That's correct, I added a fix.

Reply:
Ah right, I missed that this was an unnamed variable, thanks for clearing this up.